### PR TITLE
[eglib] Fix g_strerror / monoeg_g_strerror

### PR DESCRIPTION
**Pull request by cherusker**
```
gstr.c: In function ‘monoeg_g_strerror’:
gstr.c:238:13: warning: assignment makes integer from pointer without a cast [-Wint-conversion]
   while ((r = strerror_r (errnum, buff, buff_len - 1))) {
             ^
```

This warning caught my eye when I compiled Mono this morning. The current code is not harmful but, if I am not mistaken, it can issue invalid error messages: `g_strdup_printf ("Invalid Error code '%d'", errnum)` will be used for every `errnum` on platforms that use the "GNU-specific" version of [`strerror_r`](https://linux.die.net/man/3/strerror_r) (like my Fedora 26 machine).

### CODE REVIEW

**Comment by akoeplinger** [comment]
build

File: eglib/src/gstr.c [hunk 1]
```diff
@@ -38,6 +38,12 @@
 
 #include <errno.h>
 
+#if defined (_POSIX_C_SOURCE) && defined (_GNU_SOURCE)
+#if (_POSIX_C_SOURCE >= 200112L) && !_GNU_SOURCE
+#define HAVE_STRERROR_R_XSI
```
Comment (kumpera) [inline eglib/src/gstr.c]: HAVE_xxxx macros should only be defined by configure.

It's ok to check it here, but use a different name prefix, like USE_xxx.

Additionally, there should be a comment on why this is being checked.

File: eglib/src/gstr.c [hunk 2]
```diff
@@ -251,10 +258,16 @@ g_strerror (gint errnum)
 			error_messages [errnum] = g_strdup (buff);
 		if (buff != tmp_buff)
 			g_free (buff);
-#else
+#else /* HAVE_STRERROR_R_XSI */
+		buff = strerror_r (errnum, buff, buff_len);
```
Comment (kumpera) [inline eglib/src/gstr.c]: If we are to trust the docs, this is ok, as buff would alternatively point to an immutable buffer.

**Comment by kumpera** [comment]
There are some minor changes in the code, plus it would be nice to expand the commit message to include further explanation and pointers.

File: eglib/src/gstr.c (hunk 2 above)
Comment (cherusker) [inline eglib/src/gstr.c]: Usually, I trust the man pages - I guess that's a safe bet? In addition, it also makes a lot of sense that these strings are immutable; the custom string buffer will only be used for personalised messages.

**Comment by cherusker** [comment]
@kumpera I addressed all your thoughts - could you please have another look? :)

**Comment by kumpera** [comment]
LGTM but we need to evaluate the test failures before merging

**Comment by cherusker** [comment]
I will quickly rebase the PR (which might fix the issue with API Diff?) and will then try to rebuild all lanes until they are green; @akoeplinger just added this feature and I wanted to test it anyways :) As far as I understand, the current issues look unrelated to the changes.

**Comment by cherusker** [comment]
@monojenkins build OS X i386

**Comment by cherusker** [comment]
@monojenkins build OS X x64

**Comment by cherusker** [comment]
@monojenkins build Windows i386

**Comment by cherusker** [comment]
@monojenkins build Windows x64

**Comment by cherusker** [comment]
@monojenkins build Linux ARMv5 soft float

**Pull request merged**